10_rest/kasunfdo/pkg/puppy/rest.go

```diff
@@ -0,0 +1,98 @@
+package puppy
+
+import (
+	"net/http"
+	"strconv"
+
+	"github.com/go-chi/chi"
+	"github.com/go-chi/render"
+)
+
+// APIHandler implements REST API handlers of Puppy.
+type APIHandler struct {
+	store Storer
+}
+
+// NewAPIHandler creates a APIHandler with given storer.
+func NewAPIHandler(storer Storer) *APIHandler {
+	return &APIHandler{store: storer}
+}
+
+// HandleGetPuppyByID retrieves puppy by id.
+func (a *APIHandler) HandleGetPuppyByID(w http.ResponseWriter, r *http.Request) {
+	id, err := strconv.ParseUint(chi.URLParam(r, "id"), 10, 64)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		http.Error(w, NewError(ErrInvalid, "").Error(), http.StatusBadRequest)
+		return
+	}
+
+	puppy, err := a.store.ReadPuppy(id)
+	if err != nil {
+		w.WriteHeader(http.StatusNotFound)
+		http.Error(w, err.Error(), http.StatusNotFound)
+		return
+	}
+	render.JSON(w, r, puppy)
+}
+
+// HandlePostPuppy adds puppy into the store.
+func (a *APIHandler) HandlePostPuppy(w http.ResponseWriter, r *http.Request) {
+	var p Puppy
+	if err := render.DecodeJSON(r.Body, &p); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		http.Error(w, NewError(ErrInvalid, "").Error(), http.StatusBadRequest)
+		return
+	}
+
+	id, err := a.store.CreatePuppy(p)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	p.ID = id
+	render.Status(r, http.StatusCreated)
+	render.JSON(w, r, p)
+}
+
+// HandlePutPuppy updates puppy in the store.
+func (a *APIHandler) HandlePutPuppy(w http.ResponseWriter, r *http.Request) {
+	var p Puppy
+	if err := render.DecodeJSON(r.Body, &p); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		http.Error(w, NewError(ErrInvalid, "").Error(), http.StatusBadRequest)
+		return
+	}
+
+	if err := a.store.UpdatePuppy(p); err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	render.Status(r, http.StatusOK)
+	render.JSON(w, r, "puppy updated")
+}
+
+// HandleDeletePuppy deletes puppy in store by id.
+func (a *APIHandler) HandleDeletePuppy(w http.ResponseWriter, r *http.Request) {
+	if id, err := strconv.ParseUint(chi.URLParam(r, "id"), 10, 64); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		http.Error(w, NewError(ErrInvalid, "").Error(), http.StatusBadRequest)
+		return
+	} else if err := a.store.DeletePuppy(id); err != nil {
```

Comment:
happy path align to left, you are returning, no need for the else

Reply:
else if is used to preserve inline variables of if clause. :D